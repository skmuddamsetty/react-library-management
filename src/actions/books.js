import uuid from 'uuid';
// ADD_BOOK
export const addBook = ({
  title = '',
  description = '',
  price = 0,
  publishedAt = 0,
}) => ({
  type: 'ADD_BOOK',
  book: {
    id: uuid(),
    title,
    description,
    price,
    publishedAt,
  },
});
// EDIT_BOOK
export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates,
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id,
});
