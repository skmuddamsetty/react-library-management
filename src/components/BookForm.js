import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class BookForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.book ? props.book.title : '',
      description: props.book ? props.book.description : '',
      price: props.book ? (props.book.price / 100).toString() : '',
      createdAt: props.book ? moment(props.book.publishedAt) : moment(),
      calendarFocused: false,
      publisher: props.book ? props.book.publisher : '',
      error: '',
    };
  }

  titleChangeHandler = (e) => {
    const title = e.target.value;
    this.setState(() => ({
      title,
    }));
  };
  descriptionChangeHandler = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };
  priceChangeHandler = (e) => {
    const price = e.target.value;
    if (!price || price.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        price,
      }));
    }
  };
  dateChangeHandler = (createdAt) => {
    if (createdAt) {
      this.setState(() => {
        createdAt;
      });
    }
  };
  dateFocusChangeHandler = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused,
    }));
  };
  publisherChangeHandler = (e) => {
    const value = e.target.value;
    this.setState(() => ({ publisher: value }));
  };
  submitHandler = (e) => {
    e.preventDefault();
    if (!this.state.title || !this.state.price) {
      this.setState(() => ({
        error: 'Please provide title and price for the book!',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        price: parseFloat(this.state.price, 10) * 100, // converting to cents by multiplying with 100
        publishedAt: moment(this.state.createdAt).valueOf(), // since createdAt is a moment object converting it to timestamp using valueOf
      });
    }
  };
  render() {
    return (
      <form onSubmit={this.submitHandler} className='form'>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          className='text-input'
          type='text'
          placeholder='Title'
          autoFocus
          value={this.state.title}
          onChange={this.titleChangeHandler}
        />
        <input
          className='text-input'
          type='text'
          placeholder='Publisher'
          autoFocus
          value={this.state.publisher}
          onChange={this.publisherChangeHandler}
        />
        <input
          className='text-input'
          type='text'
          placeholder='Amount'
          value={this.state.price}
          onChange={this.priceChangeHandler}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.dateChangeHandler}
          onFocusChange={this.dateFocusChangeHandler}
          focused={this.state.calendarFocused}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className='text-area'
          placeholder='Add a description for your book'
          value={this.state.description}
          onChange={this.descriptionChangeHandler}
        ></textarea>
        <div>
          <button className='button'>Save Book</button>
        </div>
      </form>
    );
  }
}

export default BookForm;
