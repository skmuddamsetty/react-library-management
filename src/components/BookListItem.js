import React from 'react';
import { Link } from 'react-router-dom';

const BookListItem = ({
  title,
  description,
  price,
  publishedAt,
  id,
  dispatch,
  history,
}) => (
  <div>
    <Link to={`/edit-book/${id}`}>
      <h3>
        {title}
        <span> - {description}</span>
      </h3>
    </Link>
    <p>{price}</p>
    <p>{publishedAt}</p>
  </div>
);

export default BookListItem;
