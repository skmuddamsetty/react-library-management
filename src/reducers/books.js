const booksReducerDefaultState = [];

const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      return [...state, action.book];
    case 'REMOVE_BOOK':
      let books = [...state];
      books = books.filter((book) => book.id !== action.id);
      return books;
    case 'EDIT_BOOK':
      return state.map((book) => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    case 'SET_BOOKS':
      return action.books;
    default:
      return state;
  }
};

export default booksReducer;
