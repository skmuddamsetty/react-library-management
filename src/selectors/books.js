import moment from 'moment';
// Get Visible Books
const getVisibleBooks = (books, { text, sortBy, startDate, endDate }) => {
  const filteredBooks = books.filter((book) => {
    const publishedAtMoment = moment(book.publishedAt);
    const startDateMatch = startDate
      ? startDate.isSameOrBefore(publishedAtMoment, 'day')
      : true;
    const endDateMatch = endDate
      ? endDate.isSameOrAfter(publishedAtMoment, 'day')
      : true;
    const textMatch = book.title.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  });
  return filteredBooks.sort((a, b) => {
    if (sortBy === 'date') {
      return a.publishedAt < b.publishedAt ? 1 : -1; // decreasing order
    } else if (sortBy === 'price') {
      return a.price < b.price ? 1 : -1;
    }
  });
};

export default getVisibleBooks;
