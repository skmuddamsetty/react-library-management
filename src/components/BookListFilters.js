import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByPrice } from '../actions/filters';
const BookListFilters = (props) => (
  <div>
    <input
      type='text'
      value={props.filters.text}
      onChange={(e) => {
        props.dispatch(setTextFilter(e.target.value));
      }}
    />
    <select
      onChange={(e) => {
        if (e.target.value === 'date') {
          props.dispatch(sortByDate());
        } else if (e.target.value === 'price') {
          props.dispatch(sortByPrice());
        }
      }}
      value={props.filters.sortBy}
    >
      <option value='price'>Price</option>
      <option value='date'>Published Date</option>
    </select>
  </div>
);

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(BookListFilters);
