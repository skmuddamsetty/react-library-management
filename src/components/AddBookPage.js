import React from 'react';
import BookForm from './BookForm';
import { connect } from 'react-redux';
import { startAddBook } from '../actions/books';

export class AddBookPage extends React.Component {
  onSubmit = (book) => {
    // we should not use the below line as this makes testing this component harder
    // props.dispatch(addBook(book));
    // we are calling the addBook property which has been obtained from mapDispatchToProps
    this.props.startAddBook(book);
    // navigating the user to the dashboard page
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Book</h1>
          </div>
        </div>
        <div className='content-container'>
          <BookForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}
// const AddBookPage = (props) => (
//   <div>
//     <h1>Add Book Page</h1>
//     <BookForm
//       onSubmit={(book) => {
//         // we should not use the below line as this makes testing this component harder
//         // props.dispatch(addBook(book));
//         // we are calling the onSubmit property which has been obtained from mapDispatchToProps
//         props.onSubmit(book);
//         // navigating the user to the dashboard page
//         props.history.push('/');
//       }}
//     />
//   </div>
// );

// const mapDispatchToProps = (dispatch) => {
//   return {
//     onSubmit: (book) => dispatch(addBook(book)),
//   };
// };

// syntax for implicitly returning the object. this is same as the above commented code but uses a short hand syntax
const mapDispatchToProps = (dispatch) => ({
  startAddBook: (book) => dispatch(startAddBook(book)),
});
export default connect(undefined, mapDispatchToProps)(AddBookPage);
