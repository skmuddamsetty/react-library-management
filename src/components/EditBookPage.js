import React from 'react';
import { connect } from 'react-redux';
import BookForm from './BookForm';
import { editBook, removeBook } from '../actions/books';

export class EditBookPage extends React.Component {
  onSubmit = (book) => {
    this.props.editBook(this.props.book.id, book);
    this.props.history.push('/');
  };

  removeBookHandler = () => {
    this.props.removeBook({ id: this.props.book.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        Edit Book Page
        <BookForm book={this.props.book} onSubmit={this.onSubmit} />
        <button onClick={this.removeBookHandler}>Remove Book</button>
      </div>
    );
  }
}
// const EditBookPage = (props) => {
//   return (
//     <div>
//       Edit Book Page
//       <BookForm
//         book={props.book}
//         onSubmit={(book) => {
//           props.dispatch(editBook(props.book.id, book));
//           props.history.push('/');
//         }}
//       />
//       <button
//         onClick={() => {
//           props.dispatch(removeBook({ id: props.book.id }));
//           props.history.push('/');
//         }}
//       >
//         Remove Book
//       </button>
//     </div>
//   );
// };

const mapStateToProps = (state, props) => {
  return {
    book: state.books.find((book) => book.id === props.match.params.id),
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editBook: (id, book) => dispatch(id, editBook(book)),
  removeBook: ({ id }) => dispatch(removeBook({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditBookPage);
