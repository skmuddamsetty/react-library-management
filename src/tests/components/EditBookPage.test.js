import React from 'react';
import { shallow } from 'enzyme';
import { EditBookPage } from '../../components/EditBookPage';
import books from '../fixtures/books';

let editBook, removeBook, history, wrapper;

beforeEach(() => {
  editBook = jest.fn();
  removeBook = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditBookPage
      editBook={editBook}
      history={history}
      removeBook={removeBook}
      book={books[2]}
    />
  );
});

test('should render EditBookPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editBook', () => {
  wrapper.find('BookForm').prop('onSubmit')(books[2]);
  expect(history.push).toHaveBeenCalledWith('/');
  expect(editBook).toHaveBeenLastCalledWith(books[2].id, books[2]);
});

test('should handle removeBook', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenCalledWith('/');
  expect(removeBook).toHaveBeenLastCalledWith({ id: books[2].id });
});
