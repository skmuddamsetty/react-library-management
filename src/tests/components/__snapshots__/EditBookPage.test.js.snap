// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditBookPage correctly 1`] = `
<div>
  Edit Book Page
  <BookForm
    book={
      Object {
        "description": "This is the first node book",
        "id": "3",
        "price": 64500,
        "publishedAt": 345600000,
        "title": "Node Book",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Book
  </button>
</div>
`;
